### ISSUE TITLE
[question] Concurrent execution of subscription messages

### ISSUE THREAD

**Issue by nosammai** [nats-io/nats.py#272]
Hi,

I am not sure if I am understanding it incorrectly, but it seems like you can only have one outstanding callback running for when a message is consumed on a subscription. Is there a way to configure it to process more than one message at a time? Or is the right pattern to have the callback just use `asyncio.create_task()` to fire-and-forget the actual message processing, so it can return and allow another message to be grabbed off the subscription?

Example code: 
```python
import asyncio
import time

import nats

async def handler(msg):
    await asyncio.sleep(1)
    print(time.ctime(), f"Received: {msg.data.decode()}")


async def publish(n):
    for i in range(5):
        print(time.ctime(), f"Published {i}")
        await n.publish("test", f"Hello {i}!".encode())


async def main():
    n = await nats.connect(servers=["nats://localhost:4222"])
    sub = await n.subscribe("test", "", handler)
    asyncio.create_task(publish(n))
    await asyncio.sleep(10)


if __name__ == '__main__':
    asyncio.get_event_loop().run_until_complete(main())
```
Output: you can see one message is processed at a time
```
Tue Feb  8 14:00:08 2022 Published 0
Tue Feb  8 14:00:08 2022 Published 1
Tue Feb  8 14:00:08 2022 Published 2
Tue Feb  8 14:00:08 2022 Published 3
Tue Feb  8 14:00:08 2022 Published 4
Tue Feb  8 14:00:09 2022 Received: Hello 0!
Tue Feb  8 14:00:10 2022 Received: Hello 1!
Tue Feb  8 14:00:11 2022 Received: Hello 2!
Tue Feb  8 14:00:12 2022 Received: Hello 3!
Tue Feb  8 14:00:13 2022 Received: Hello 4!
```

**Comment by wallyqs** [comment]
There is no way to configure concurrency at the moment, each subscription would have its own task so that they can process messages in parallel but within each subscription there is head of line blocking.  I made the example below that creates a `max_tasks` for a subscription so that you can have a bounded number of inflight messages being processed by a single subscription:

```python
import asyncio
import time

import nats

class Component:
    def __init__(self, nc):
        self._nc = nc
        self._done = asyncio.Future()

    async def run_forever(self):
        await self._done

    async def stop(self):
        if self._done:
            self._done.cancel()

    async def subscribe(self, *args, **kwargs):
        if 'max_tasks' in kwargs:
            sem = asyncio.Semaphore(kwargs['max_tasks'])
            callback = kwargs['cb']

            async def release(msg):
                try:
                    await callback(msg)
                finally:
                    sem.release()

            async def cb(msg):
                await sem.acquire()
                asyncio.create_task(release(msg))

            kwargs['cb'] = cb
      
            # Add some concurrency to the handler.
            del kwargs['max_tasks']

        return await self._nc.subscribe(*args, **kwargs)

    async def publish(self, *args):
        await self._nc.publish(*args)

async def main():
    nc = await nats.connect(servers=["nats://localhost:4222"])   
    c = Component(nc)

    async def handler(msg):
        # Cause some head of line blocking
        await asyncio.sleep(0.5)
        print(time.ctime(), f"Received on 'test': {msg.data.decode()}")
      
    async def wildcard_handler(msg):
        # Cause some head of line blocking
        await asyncio.sleep(1)
        print(time.ctime(), f"Received on '>': {msg.data.decode()}")

    await c.subscribe("test", cb=handler, max_tasks=3)
    await c.subscribe(">", cb=wildcard_handler)

    async def producer(label):
        i = 0
        while True:
            # print(time.ctime(), f"Published {i}")
            await c.publish("test", f"{label} :: Hello {i}!".encode())
            await asyncio.sleep(0.2)
            i += 1

    asyncio.create_task(producer("A"))
    asyncio.create_task(producer("B"))
    asyncio.create_task(producer("C"))

    await c.run_forever()

if __name__ == '__main__':
    asyncio.run(main())
```

Sample Output:

```
Wed Feb  9 15:06:38 2022 Received on 'test': A :: Hello 0!
Wed Feb  9 15:06:38 2022 Received on 'test': B :: Hello 0!
Wed Feb  9 15:06:38 2022 Received on 'test': C :: Hello 0!
Wed Feb  9 15:06:38 2022 Received on '>': A :: Hello 0!
Wed Feb  9 15:06:38 2022 Received on 'test': A :: Hello 1!
Wed Feb  9 15:06:38 2022 Received on 'test': B :: Hello 1!
Wed Feb  9 15:06:38 2022 Received on 'test': C :: Hello 1!
Wed Feb  9 15:06:39 2022 Received on 'test': A :: Hello 2!
Wed Feb  9 15:06:39 2022 Received on 'test': B :: Hello 2!
Wed Feb  9 15:06:39 2022 Received on 'test': C :: Hello 2!
Wed Feb  9 15:06:39 2022 Received on '>': B :: Hello 0!
Wed Feb  9 15:06:39 2022 Received on 'test': A :: Hello 3!
Wed Feb  9 15:06:39 2022 Received on 'test': B :: Hello 3!
Wed Feb  9 15:06:39 2022 Received on 'test': C :: Hello 3!
```
